lib/mongo/retryable/base_worker.rb

```diff
@@ -0,0 +1,90 @@
+# frozen_string_literal: true
+# encoding: utf-8
+
+# Copyright (C) 2015-2020 MongoDB Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module Mongo
+  module Retryable
+
+    # The abstract superclass for workers employed by Mongo::Retryable.
+    #
+    # @api private
+    #
+    # @since 2.19.0
+    class BaseWorker
+      extend Forwardable
+
+      def_delegators :retryable,
+        :client,
+        :cluster,
+        :select_server
+
+      # @return [ Mongo::Retryable ] retryable A reference to the client object
+      #   that instatiated this worker.
+      attr_reader :retryable
+
+      # Constructs a new worker.
+      #
+      # @example Instantiating a new read worker
+      #   worker = Mongo::Retryable::ReadWorker.new(self)
+      #
+      # @example Instantiating a new write worker
+      #   worker = Mongo::Retryable::WriteWorker.new(self)
+      #
+      # @param [ Mongo::Retryable ] retryable The client object that is using
+      #   this worker to perform a retryable operation
+      def initialize(retryable)
+        @retryable = retryable
+      end
+
+      private
+
+      # Returns a list of exception classes that are generally retryable. 
+      def retryable_exceptions
+        [
+          Error::ConnectionPerished,
+          Error::ServerNotUsable,
+          Error::SocketError,
+          Error::SocketTimeoutError
+        ].freeze
+      end
+
+      # Tests to see if the given exception instance is of a type that can
+      # be retried.
+      #
+      # @return [ Boolean ] true if the exception is retryable.
```

Comment:
```suggestion
      # @return [ true | false ] true if the exception is retryable.
```
We agreed not to use `Boolean` since such type does not exist in Ruby.

However, YARD (as well as Sorbet and RBS) allows and encourages using `Boolean`, so we may want to re-discuss this topic.

Reply:
I've been writing Ruby code for more than 20 years, and I *still* forget that there's no Boolean superclass in the language. :)